### ISSUE TITLE
Increased latency over time in V49

### ISSUE THREAD

**Issue by Andrews54757** [pyamsoft/tetherfi#354]
Since updating the app (via play store) to V49, I've noticed an increased latency when visiting websites, manifesting as long loading times. This latency appears to grow with time and usage, only resolving by restarting the app. I've since downgraded to V47 and I have not seen this issue again.

Thanks for the amazing app!

### Device information

Galaxy S10, android v12.

### TetherFI configuration

5Ghz direct wifi connection, with always alive on and max performance power balance settings. unrestricted power management.

**Comment by pyamsoft** [comment]
Interesting, while the structure of the code has changed the underlying logic is relatively untouched.

If latency gets worse as you browse longer, I wonder if this is due to bigger and bigger copy actions from the data and bandwidth tracker.

I may have to remove the feature for now if it is affecting speeds

**Comment by pyamsoft** [comment]
About how long does it take to occur, and what websites are you browsing generally when you notice?

**Comment by Andrews54757** [comment]
It tends to occur within 10 minutes of browsing. affects websites from google to youtube.

**Comment by pyamsoft** [comment]
Thank you

Let me set up some tests to see if I can reproduce the problem

**Comment by pyamsoft** [comment]
I've confirmed some kind of performance regression from 47 to 49, but it's deep in there - it will be a little tricky. I do not have a good chunk of time this week to sit down, but I suspect it has to do with client handling.

Assuming I can fix this, I will build more tests that ensure performance stays snappy for future versions, thanks :)

**Comment by pyamsoft** [comment]
Hello

I've noticed a few things regarding the new version 49 (warning: LONG)

1. TetherFi is powered by a nice Kotlin library called `KTOR` made by a company called Jetbrains. KTOR used to be at version 2.X, which we were using internally. This library made internet "connect" calls that took, lets say, around 100ms on a test machine. TetherFi from version 49 onward uses KTOR version 3 which just released, and this version appears to connect every so slightly slower, at 150ms.

2. TetherFi version 49 deprecated the power balance setting, which differs from the old version in one significant way. The absolute fastest performance was on unrestricted balance mode, but changes in 49 deprecated this mode and removed it from the codebase. This change was made without me knowing all of the facts, and I believe this is a cause of significant slowdown. To keep a complex topic simple, after you device makes "enough" connections, where "enough" is determined by the "number of CPU cores you have", the connection speed will begin to slow down as internal we are fighting for resources.

3. This fight for resources was compounded by the fact that TetherFi used to never close socket connections, which meant that old "dead" Internet connections just sat around doing nothing and making things slow. Version 49 changed this to enforce a "dead connection cleanup" every 1 minute but this timeout period was obviously too slow. Version 50 will reduce this time down to 10 seconds.

Given these three changes, I wrote a simple shell script called `stresstest` that lives in the `/bin` folder in this repository, where I make 100 curl requests to `example.com` and time each one.
Before making these adjustments, 100 requests would take around 40 seconds on a Galaxy S24 - after making this change with unrestricted mode it takes around 10 seconds (depending on your internet speed)

I believe these changes will significantly improve the experience for you. Version 50 will be entering the store Beta hopefully sometime tonight or tomorrow and I will release it live once testing is completed.

Thank you for your patience and your contribution to the project
